codec/src/buffer.rs

```diff
@@ -0,0 +1,275 @@
+//! Buffer implementation with advanced safety features
+
+use crate::{error::Error, varint};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+/// A buffer for reading codec data with safety checks
+#[derive(Debug)]
+pub struct ReadBuffer {
+    /// The underlying bytes
+    inner: Bytes,
+}
+
+impl ReadBuffer {
+    /// Creates a new reader from bytes
+    pub fn new(bytes: Bytes) -> Self {
+        Self { inner: bytes }
+    }
+
+    /// Reads a varint-encoded unsigned integer
+    #[inline]
+    pub fn read_varint(&mut self) -> Result<u64, Error> {
+        varint::decode_varint(&mut self.inner)
+    }
+
+    /// Gets a byte from the buffer
+    #[inline]
+    pub fn get_u8(&mut self) -> Result<u8, Error> {
+        if !self.has_remaining() {
+            return Err(Error::EndOfBuffer);
+        }
+        Ok(self.inner.get_u8())
+    }
+
+    /// Gets remaining bytes
+    #[inline]
+    pub fn remaining(&self) -> usize {
+        self.inner.remaining()
+    }
+
+    /// Checks if the buffer has any remaining bytes
+    #[inline]
+    pub fn has_remaining(&self) -> bool {
+        self.remaining() > 0
+    }
+
+    /// Ensures the buffer has at least `size` bytes remaining
+    #[inline]
+    pub fn ensure_size(&self, size: usize) -> Result<(), Error> {
+        if self.remaining() < size {
+            return Err(Error::EndOfBuffer);
+        }
+        Ok(())
+    }
+
+    /// Advance the buffer by `cnt` bytes
+    #[inline]
+    pub fn advance(&mut self, cnt: usize) {
+        self.inner.advance(cnt);
+    }
+
+    /// Returns a slice of the remaining bytes in the buffer
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
```

Comment:
`chunk()` is not guaranteed to return a slice with all remaining: https://docs.rs/bytes/1.10.1/bytes/struct.Bytes.html#method.chunk

I had to do this to get around its semantics: https://github.com/commonwarexyz/monorepo/blob/5d07654087db22919e7316345f93f05c305b69d5/utils/src/array/mod.rs#L62-L71

Reply:
Just got rid of the `chunk()` functions since they are n't used